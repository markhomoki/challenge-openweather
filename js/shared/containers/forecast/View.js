// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import { fetchData, importCSV } from 'app/actions';
import type { Dispatch, ImportDataType } from 'app/types';
import { CSVUploader, LocationSelector } from 'app/components';
import type { ForecastCollectionType } from './types';

type Props = {
	dispatch: Dispatch,
	forecast: ForecastCollectionType,
};

type State = {
	currentLocation: string,
};

@connect((state) => {
	return {
		forecast: state.forecast,
	};
})
export default class View extends React.PureComponent<Props, State> {

	times: Array<string> = [
		'00:00',
		'03:00',
		'06:00',
		'09:00',
		'12:00',
		'15:00',
		'18:00',
		'21:00',
	];

	state = {
		currentLocation: 'London,GB',
	};

	componentDidMount() {
		this.fetchData();
	}

	fetchData = () => {
		this.props.dispatch(fetchData({
			type: 'FORECAST',
			url: `/forecast?q=${this.state.currentLocation}`,
		}));
	}

	handleImport = (data: ImportDataType) => {
		this.props.dispatch(importCSV(data));
	}

	handleLocationChange = (currentLocation: string) => {
		this.setState({ currentLocation }, this.fetchData);
	}

	renderContent = () => {
		const {
			forecast,
		} = this.props;

		if (forecast.isLoading) {
			return <p className="loader">Loading...</p>;
		}

		return (
			<table>
				<thead>
					<tr>
						<th />
						{this.times.map(time => <th key={time}>{time}</th>)}
					</tr>
				</thead>
				<tbody>
					{forecast.data.map((r) => {
						const dates = r.date.split('-');
						return (
							<tr key={r.date}>
								<th>
									<span className="hide-on-mobile">{dates[0]}-</span>
									<span>{dates[1]}-</span>
									<span>{dates[2]}</span>
								</th>
								{this.times.map((time) => {
									const temp = r.times[time] ? r.times[time].celsius : null;
									return <td key={time}>{temp ? <span>{temp}<span className="hide-on-mobile"> Â°C</span></span> : ''}</td>;
								})}
							</tr>
						);
					})}
				</tbody>
			</table>
		);
	}

	render() {

		return (
			<div className="site-inner">
				<div className="forecast-actions">
					<div className="select-city-wrapper">
						<LocationSelector onFetch={this.handleLocationChange} />
					</div>
					<span className="or">OR</span>
					<div className="csv-import-wrapper">
						<CSVUploader onImport={this.handleImport} />
					</div>
				</div>

				<h1>Weather Forecast</h1>
				{this.renderContent()}
			</div>
		);
	}

}
