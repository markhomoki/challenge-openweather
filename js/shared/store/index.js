// @flow

import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import reducers from 'app/reducers';

const middleware = process.env.NODE_ENV === 'production' ? [thunk] : [thunk, createLogger()];

export default createStore(
	reducers,
	applyMiddleware(...middleware),
);
