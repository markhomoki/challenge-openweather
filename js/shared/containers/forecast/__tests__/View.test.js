import React from 'react';
import renderer from 'react-test-renderer';
import store from 'app/store';
import View from '../View';

function sleep(ms) {
	return new Promise(resolve => setTimeout(resolve, ms));
}

describe('View', () => {

	let wrapper;

	beforeEach(() => {
		wrapper = shallow(
			<View store={store} />
		);
	});

	it('renders correctly', () => {
		const tree = renderer.create(<View store={store} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('shows loader until fetching then hides loader', async () => {
		expect(wrapper.dive().contains(<p className="loader">Loading...</p>)).toEqual(true);
		wrapper.dive().instance().fetchData();
		await sleep(1000);
		expect(wrapper.dive().contains(<p className="loader">Loading...</p>)).toEqual(false);
	});
});
