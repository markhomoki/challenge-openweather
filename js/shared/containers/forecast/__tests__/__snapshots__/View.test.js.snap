// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View renders correctly 1`] = `
<div
  className="site-inner"
>
  <div
    className="forecast-actions"
  >
    <div
      className="select-city-wrapper"
    >
      <div
        className="form-field-select"
      >
        <span
          className="form-field-label"
        >
          Select Location
        </span>
        <select
          onChange={[Function]}
          value="London,GB"
        >
          <option
            value="London,GB"
          >
            London
          </option>
          <option
            value="Paris,FR"
          >
            Paris
          </option>
          <option
            value="Barcelona,ES"
          >
            Barcelona
          </option>
          <option
            value="Rome,IT"
          >
            Rome
          </option>
        </select>
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          Fetch
        </button>
      </div>
    </div>
    <span
      className="or"
    >
      OR
    </span>
    <div
      className="csv-import-wrapper"
    >
      <div
        className="form-field-file-picker"
      >
        <span
          className="form-field-label"
        >
          Select File
        </span>
        <input
          accept=".csv"
          onChange={[Function]}
          type="file"
        />
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          Import
        </button>
      </div>
    </div>
  </div>
  <h1>
    Weather Forecast
  </h1>
  <p
    className="loader"
  >
    Loading...
  </p>
</div>
`;
